### PR TITLE
Add section for Promises (JavaScript)

### PR DESCRIPTION
**Pull request by danielbrzn**
This PR aims to add a section regarding Promises in Javascript.

### CODE REVIEW

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: I wouldn't call it a "method call". perhaps you can say that an operation is executed asynchronously and its result may not be obtained immediately as a function return value would.

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: a diagram that shows the execution timeline would be nice to have.

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: good example. a simpler example would actually be timer, i.e. `setTimeout()`. You can explain what is callback straight up here in a setTimeout example and then talk about async execution and event loop

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: "it can be difficult to predict how" - I think maybe it's better to not describe as "predict" (as though it's a random variable). Perhaps we can say that retrievePrice is non-blocking (this is keyword) and the I/O operation would be done in a different thread (depends on platform implementation) etc.

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: as mentioned before, an explanation of callback-based programming first before introducing Promises can help with this.

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: avoid linking to external resources that may become unavailable sometime later. if need be and allowed, put the image in the repository too!

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: titles can be more elaborated and to the point, perhaps here we can say "How to use Promises"

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md:187]: there are other libraries that expand on the native Promise library provided by default. I personally use Bluebird where they have the .spread() functionality. check them out and see if you're able to add in to the doc

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: It'd be good to mention that Promises are available from which versions of JavaScript and on which platforms.

File: contents/javascript/Javascript-promises.md
Comment (mauris) [inline contents/javascript/Javascript-promises.md]: use `{` and `}` even if it's one line inside the if-branch.

File: contents/javascript/Javascript-promises.md
Comment (junkiattan) [inline contents/javascript/Javascript-promises.md:4]: You can add a Context title to let readers know the context of Promises in Javascript

File: contents/javascript/Javascript-promises.md
Comment (junkiattan) [inline contents/javascript/Javascript-promises.md]: I feel the flow would be better if you had put the Benefits section before the Other Functions section. As a reader, I would want to know the implementation details first to judge the usability before understanding the benefits to further solidifying the point that Promises are good

File: contents/javascript/Javascript-promises.md
Comment (junkiattan) [inline contents/javascript/Javascript-promises.md]: Earlier in your text, you said that Promises have 3 states. Do be careful not to confuse readers. Perhaps you were referring to "2 completed states" instead.

File: contents/javascript/Javascript-promises.md
Comment (junkiattan) [inline contents/javascript/Javascript-promises.md]: In general for all your code examples, is it possible to color code them like in actual code? I have seen some of the resources with code examples that do this. 

This makes the code stand out more, and helps in the flow.

File: contents/javascript/Javascript-promises.md
Comment (junkiattan) [reply contents/javascript/Javascript-promises.md]: The example is good and easy to understand for me

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md]: Managed to do it with Markdown code syntax highlighting. Thanks for the feedback!

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md]: Good catch.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md:4]: Do you have an example for this? I'm not quite sure what you mean.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md]: Good catch, thanks!

File: contents/javascript/Javascript-promises.md
Comment (junkiattan) [reply contents/javascript/Javascript-promises.md:4]: What I meant was to have a Context title, like your Introduction title or How to use Promises title.

That way people would instantly be mentally prepped of what to anticipate for your first section. This is a small change and completely optional.

**Comment by danielbrzn** [comment]
Hi Prof @damithc,

May I get a review on this PR?

Thanks!

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: This text up to this point probably need another go. Start with a simple example and gently explain what synchronous/asynchronous means and build up until the reader understands the problem context. As @junkiattan mentioned, you can give a heading to the section too. e.g., **Why Promises?**

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: If you name the previous section `Why Promises?`, you can name this `What is a Promise?`

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: The first sentence hints that promises can be understood by relating to a similar concept in real life; yet the second sentence says something that is hardly related to real life. On a similar note, what is meant by '_resolve_ a promise'. That's not something a casual reader can be expected to know.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Given that this resource is not meant to _teach_ someone how to use promises, I wonder if we should omit this type of nitty-gritty details.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: May be reframe this as 'How Promises are used' and give some nice examples of how Promises are used? The current title suggests that the section teaches how to use Promises. Normally our learning resources don't try to teach topics. Rather, we want to motivate the reader to learn the topic and point to other resources for the actual learning of the topic.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: This is where you explain what a callback is, but I remember seeing the term callback near the beginning of this chapter. :-p

**Review by damithc — CHANGES_REQUESTED** [review]
Some changes suggested. Read the current from a 'knowledgeable outsider' POV and you'll realize it's not easy to follow. Probably you don't see it because you already know the topic. Read a few other similar articles on Promises like [this one](https://developers.google.com/web/fundamentals/primers/promises) and try to come up with a shorter version that combines the best of those articles (and point to the original articles for a longer version).

**Comment by danielbrzn** [comment]
@damithc may I get another review? I made some changes for the content to be more user-friendly, and also removed the section about how to use them, replacing it with an "how they are used" section instead.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: This is not specific to JavaScript.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: The description up to this point can be tweaked slightly along this flow:
* Usually, ops are synchronous i.e., wait for completion before proceeding
* But some ops take a while to finish. e.g., http request 
* Doing such ops in synchronous manner is not good for user experience
* In such cases, we can use async ops. they don't wait for completion.
* but code of async ops are harder to understand/debug than sync ops

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: have you explained what a callback is?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Best explain it now rather than .. ahem ... "promise" to explain it later. 
In any case, this example is supposed to show why async code is hard to read/debug (following from the previous paragraph) but doesn't seem to do so.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: What is meant by `resolved`?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Well, this code looks harder to read than the setTimeout() code, which contradicts your claim about Promises :-p

Comment (damithc) [reply contents/javascript/Javascript-promises.md]: explain what's so good about this code.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md:183]: This code looks simpler than the httprequest example. May be use this as the running example for the entire chapter? First show how it is without Promises and then show how Promises makes it better. It's easier to follow when you use the same example throughout. e.g.,

Add two numbers locally : this can be a snyc op
Add two numbers remotely: this takes time. Need to be async.
Let's use callbacks. Here is the code.
What if we need to chain a few remote add ops? Here is the code. Oops! callback hell.
Promises to the rescue. 
Here's how they work:
First we need to create a promise. Here is the code. 
Next we need to consume the promise. Here is the code. 
Here is how we can chain. See how much better it is from callbacks.
There are some more advance stuff you can do with promises: x, y, z
There are some alternatives to promises: a, b, c.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Both images in this PR used?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md:1]: Better than before 👍 Some way to go before you can make it understandable to a 'knowledgeable outsider' (not only to existing JS programmers). I suspect you can some inspiration from this article https://scotch.io/tutorials/javascript-promises-for-dummies (remember to acknowledge it).

**Review by damithc — CHANGES_REQUESTED** [review]
Good progress. Need a few more iterations though, as this is not an easy topic to make accessible to a 'knowledgeable outsider'.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md]: Yes it's been explained in the paragraph before the code snippet was shown.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md]: I think it touches on why async code is hard to debug as I explain that the program execution isn't what the user expected, e.g. wait for 1s then log with the updated message.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md:1]: Added a snippet to acknowledge their example.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md]: Removed.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md:183]: I feel like using the SetTimeout example would be easier for new users of JS as it's something they can try out on their own in Chrome/IDE to see how it actually works, and how we can resolve the problems of async operations using Promises. 

This file reading example is just a way of showing how you can still read files fast using async ops that allow it to be read in the background.

**Comment by danielbrzn** [comment]
@damithc, may I get another review?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: This statement sounds like "promises provide a way to do things synchronously" (because it says that the rest of the program will not execute until the async code is executed). I don't know promises myself, but the statement didn't sound right to me. Did you mean "promises allows you to specify how the execution of _some part_ of the code to depend on the status of an asynchronous code" or something of the sort.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: hmm... it's unfortunate that you need to bring node into this, given that we consider node to be a separate topic altogether. Is there any way to avoid it?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Better not use first person (`I`) as this chapter might have more than one author in future.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Web scraping is another topic. Do we need to mention it here?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: While this chapter is not meant to to teach promises syntax, I think you still need to walk the reader through a small hello world type example for the reader to get some sense of what promises are. May be explain the promises basics using how to write the timer example using promises?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Image link is broken. In any case, you don't need it as you already have a similar example at the beginning of the chapter.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: You can mention callback hell here?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: setTimeout is an async function? It wasn't obvious to me at first. It is something a relatively new JS programmer supposed to know? If yes, can leave as it is. If no, explain it a bit.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Where is this `readFile()` used in the text below?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: This section and the next section seem thin. One option is to get rid of them and absorb the content into the previous section. i.e., explain them as different 'ways of using promises' and mention the benefits at the same time. For example, the benefit of avoiding callback hell can be mentioned in the timeout example.

**Review by damithc — CHANGES_REQUESTED** [review]
Getting better; a few more kinks to work out.

File: contents/javascript/Javascript-promises.md
Comment (acjh) [reply contents/javascript/Javascript-promises.md]: @damithc It's used in the "Disk I/O" section.

---
@danielbrzn 

> This is an asynchronous function [that] takes in 3 parameters, a callback and a time (in milliseconds) to wait before the callback is executed.

That sounds like 5 parameters in total. The 3 parameters are [path[, options], callback](https://nodejs.org/api/fs.html#fs_fs_readfile_path_options_callback).

Comment (damithc) [reply contents/javascript/Javascript-promises.md]: Just like we should define variables as closer to where they are used as possible, introduce terms closer to where they are used :-)

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md]: Would 1 Node.js example (file I/O) and 1 Javascript example be a good compromise? Or should we try to make both examples pure JS.

Comment (damithc) [reply contents/javascript/Javascript-promises.md]: I'll leave it to you to decide. Try to minimize dependency on node.js so that even someone who has no knowledge of node.js can understand this content, unless there is a legit reason for promises to depend on node.js

**Comment by danielbrzn** [comment]
@damithc , could I get another review :)

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md:150]: May be show how this is used too, to be consistent with other examples you have given?

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: Can remove this statement without any loss.

File: contents/javascript/Javascript-promises.md
Comment (damithc) [inline contents/javascript/Javascript-promises.md]: `an HTTP`? https://www.quora.com/A-http-or-an-http-which-is-the-correct-grammatical-usage

**Review by damithc — COMMENTED** [review]
Looks good. Just a few nits.

File: contents/javascript/Javascript-promises.md
Comment (danielbrzn) [reply contents/javascript/Javascript-promises.md:150]: Added a snippet to show how the function is called and handled.

**Comment by danielbrzn** [comment]
I've made the necessary changes.

**Comment by danielbrzn** [comment]
@damithc Are there any more changes that I should make for this chapter?

**Review by damithc — APPROVED** [review]

**Comment by damithc** [comment]
Merged. It's in good shape now. Good job doing all those revisions along the way. 👍

**Comment by danielbrzn** [comment]
Thank you for the guidance along the way 👍

**Pull request merged**
